packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -157,7 +183,7 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   /// Return the engine layer for retained rendering. When there's no
   /// corresponding engine layer, null is returned.
   @protected
-  ui.EngineLayer addToScene(ui.SceneBuilder builder, [ Offset layerOffset = Offset.zero ]);
+  void addToScene(ui.SceneBuilder builder, [ Offset layerOffset = Offset.zero ]);
```

Comment:
I reverted the change. I only changed it because it didn't feel right. It puts the responsibility of setting the `_engineLayer` field on the _caller_ of `addToScene`. This means that the caller always has to do this:

```
childLayer._engineLayer = childLayer.addToScene(builder);
```

There are two issues with this. One is that `_engineLayer` is private and so a custom layer implementation won't be able to do this correctly. The second is that the caller needs to _remember_ to do this. They also need to make sure they do this correctly. I found a bug in `_ProxyLayer` that returned another layer's engine layer, which is wrong. It only manifested after my change because my change exercises the new assertions I added in the `SceneBuilder` API.

However, there's no rush to fix this. We can do this as a follow-up PR.

Reply:
I remembered that we've had this conversation before. Maybe let's create a Github issue to track our discussions and decisions so we won't forget it again in a few months :-)